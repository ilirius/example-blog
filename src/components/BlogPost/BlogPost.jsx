import React from "react";
import LazyLoad from "react-lazyload";
import PropTypes from "prop-types";

import "./style.static.scss";

export const BlogPost = ({ post, clickTag }) => {
  const onClick = evt => {
    evt.preventDefault();
    clickTag(post.getCategory());
  };

  const onImgError = evt => {
    const { target } = evt;
    target.src = post.getBannerUrl();
    target.style.position = "static";
  };

  return (
    <article
      className="post"
      itemProp="blogPosts"
      itemScope
      itemType="http://schema.org/BlogPosting"
    >
      <meta itemProp="author" content="EuroAuto" />
      {/* <div
        className="hidden"
        itemProp="publisher"
        itemScope
        itemType="https://schema.org/Organization"
      >
        <meta itemProp="name" content="ЕвроАвто" />
        <meta itemProp="address" content="Санкт-Петербург, Левашовский пр., 15" />
        <meta itemProp="telephone" content="+78127031515" />
        <link itemProp="url" href={`https://${window.__host_main__}`} />
        <div itemProp="logo" itemScope itemType="https://schema.org/ImageObject">
          <img itemProp="url image" src="/images/logo-black.svg" />
        </div>
      </div> */}
      <meta
        itemScope
        itemProp="mainEntityOfPage"
        itemType="https://schema.org/WebPage"
        itemID={post.getUrn()}
      />
      <meta itemProp="image" content={post.getBannerUrl()} />
      <a key={post.getUrn()} href={post.getUrn()}>
        <div className="img">
          <LazyLoad height={127}>
            <img
              src={post.getBannerUrl() + "?thumbnail=350x190"}
              alt={post.getH1()}
              onError={onImgError}
            />
          </LazyLoad>
        </div>
        <div className="fx content">
          <div className="content_title" itemProp="headline">
            {post.getH1()}
          </div>
          <meta itemProp="description" content={post.getDescription()} />
          <div className="content_text" itemProp="abstract">
            {post.getPreviewText()}
          </div>
          <div className="fx content_footer">
            {/* "Y-m-d" */}
            <div className="content_footer__data" itemType="https://schema.org/dateCreated">
              <meta itemProp="datePublished" content={post.getCreatedAt("yyyy-MM-dd")} />
              <meta itemProp="dateModified" content={post.getCreatedAt("yyyy-MM-dd")} />
              {post.getCreatedAt()}
            </div>
            {post.getCategory().getId() && (
              <button className="footer__tag content_footer__tag" onClick={onClick}>
                {post.getCategory().getLabel()}
              </button>
            )}
          </div>
        </div>
      </a>
    </article>
  );
};

BlogPost.propTypes = {
  post: PropTypes.any,
  clickTag: PropTypes.func,
};

export default BlogPost;
