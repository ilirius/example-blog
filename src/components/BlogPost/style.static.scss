#blog_root {
	.post {
		border-radius: 4px;
		box-shadow: 0px 2px 16px 0px rgba(50, 50, 50, 0.14);
		cursor: pointer;
		margin: 0 15px 15px 0;
		transition: 0.22s;
		max-width: 365px;
		width: calc(50% - 15px);

		a {
			text-decoration: none;
		}

		@media (max-width: 992px) {
			& {
				margin: 0 15px 15px 0;
			}
		}

		@media (max-width: 640px) {
			& {
				width: auto;
				margin: 0 0 15px 0;
			}
		}

		&:hover {
			box-shadow: 0px 4px 24px 0 rgba(50, 50, 50, 0.24);
			text-decoration: none;

			& .content {
				&_title {
					color: #3379b7;
				}
			}
		}

		& .img {
			border-radius: 4px 4px 0 0;
			display: block;
			height: 130px;
			width: 100%;
			overflow: hidden;
			@media (min-width: 640px) and (max-width: 840px) {
				& {
					height: 110px;
				}
			}

			img {
				border-radius: 4px 4px 0 0;
				height: auto;
				width: 100%;
				object-fit: cover;
				position: relative;
				top: -35px;
			}
		}

		& .content {
			flex-direction: column;
			height: calc(100% - 130px);
			padding: 30px 30px 20px 30px;

			&_title {
				font-size: 19px;
				font-weight: 700;
				color: #333333;
				margin-bottom: 10px;
			}

			&_text {
				color: #999;
				flex: 1 1 auto;
				font-size: 13px;
				line-height: 20px;
				margin-bottom: 20px;
			}

			&_footer {
				width: 100%;
				align-items: center;

				&__data {
					color: #999;
					font-size: 11px;
					letter-spacing: 0.66px;
				}

				&__tag {
					background-color: #fff;
					border-radius: 50px;
					border: 1px solid rgba(51, 122, 183, 0.45);
					color: #337ab7;
					font-size: 9px;
					letter-spacing: 0.66px;
					line-height: 23px;
					padding: 0 11px;
					text-transform: uppercase;

					&:hover {
						background-color: #337ab7;
						border-color: #337ab7;
						color: #ffffff;
					}
				}
			}
		}
	}
}
