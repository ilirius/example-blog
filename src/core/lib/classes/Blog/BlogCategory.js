export default class BlogCategory {
  _data = {};
  _id = 0;
  constructor(id, data) {
    this._id = id;
    this._data = data;
  }

  getId() {
    return this._id;
  }

  getLabel() {
    return this.__get_data_value__("label");
  }

  getUrn() {
    return this.__get_data_value__("urn");
  }

  getDescription() {
    return this.__get_data_value__("description");
  }

  /**
   *
   *
   * @returns
   * @memberof BlogCategory
   */
  getPostsCount() {
    return this.__get_data_value__("postsCount");
  }

  /**
   * Возвращает значение из внутреннего массива данных
   * @param {string} $key
   * @param {any} $default
   * @return {any}
   */
  __get_data_value__($key, $default = null) {
    return this._data[$key] ? this._data[$key] : $default;
  }
}
