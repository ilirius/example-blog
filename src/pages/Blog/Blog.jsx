import React, { useState, useEffect, useRef } from "react";
import Helmet from "react-helmet";
import { useSelector, useDispatch } from "react-redux";
import querystring from "querystring";
import { useHistory, useLocation, useRouteMatch, useParams } from "react-router-dom";

import { PageHeader, PaginationBlock, Loader } from "../../common";
import { FilterPanel, BlogPost } from "../../components";

import {
  appBlogPagePostsRequest,
  blogPostFilterCategoriesRequest,
  appBlogPagePostsByFilterRequest,
} from "../../core/redux/blog/actions";
import {
  getBlogPosts,
  getBlogPostsPagesCount,
  getBlogPostsFilterCategories,
  getBlogPostsIsFetched,
  getBlogPostsIsFetching,
} from "../../core/redux/blog/reducers";

import "./style.static.scss";

const COUNT_PER_PAGE = 15;

export const BlogPage = () => {
  const history = useHistory();
  let { search } = useLocation();
  search = querystring.parse(search.replace("?", ""));
  const { category } = useParams();

  const match = useRouteMatch();
  const dispatch = useDispatch();

  const [currentPage, setCurrentPage] = useState(1);
  const [replace, setReplace] = useState(false);

  const posts = useSelector(state => getBlogPosts(state));
  const postsPagesCount = useSelector(state => getBlogPostsPagesCount(state));
  const filterCategories = useSelector(state => getBlogPostsFilterCategories(state));
  const isFetched = useSelector(state => getBlogPostsIsFetched(state));
  const isFetching = useSelector(state => getBlogPostsIsFetching(state));

  const prevCategory = usePrevious(category);
  const prevQuerySearch = usePrevious(search);

  let activeCategory = false;

  useEffect(() => {
    if (typeof search["_escaped_fragment_"] === "undefined") {
      dispatch(blogPostFilterCategoriesRequest());
      if (search["page"] !== void 0) {
        setCurrentPage(search["page"]);
      }

      if (category !== void 0) {
        dispatch(
          appBlogPagePostsByFilterRequest({
            category,
            replace: true,
            currentPage,
          })
        );
      } else {
        dispatch(appBlogPagePostsRequest({ currentPage }));
      }

      history.push(location.pathname + "" + location.search);
    }
  }, []);

  useEffect(() => {
    if (
      prevCategory !== category ||
      (prevQuerySearch !== void 0 && prevQuerySearch["page"] !== search["page"])
    ) {
      let _currentPage = 1;
      let _replace = replace;

      if (
        search["page"] !== void 0 ||
        (prevQuerySearch !== void 0 && prevQuerySearch["page"] !== search["page"])
      ) {
        _currentPage = parseInt(search["page"], 10);
      }

      setCurrentPage(_currentPage);

      if (
        prevCategory !== category ||
        replace ||
        history.action === "POP" ||
        typeof search["page"] === "undefined"
      ) {
        _replace = true;
      }

      if (category !== void 0) {
        dispatch(
          appBlogPagePostsByFilterRequest({
            category,
            replace: _replace,
            currentPage: _currentPage,
          })
        );
      } else {
        dispatch(appBlogPagePostsRequest({ currentPage: _currentPage, replace: _replace }));
      }
      setReplace(false);
    }
  }, [category, search]);

  /**
   * Генерация текста тега title
   * 
   * @returns {String}
   */
  const getTitle = () => {
    let title = "";

    if (isFetched) {
      if (typeof category !== "undefined" && filterCategories.length > 0) {
        let _category = filterCategories.filter(x => x.getUrn() === category);

        if (_category[0]) {
          title = _category[0].getLabel();
        }
      } else {
        title = "Полезные статьи об автомобилях для автолюбителей";
      }

      title += " | Авто Блог Евроавто";

      if (currentPage > 1) {
        title += ` | Страница ${currentPage}`;
      }
    }
    return title;
  };

  /**
   *  Генерация текста мета тега description
   * 
   * @returns {String}
   */
  const getDescription = () => {
    let description = "";

    if (isFetched) {
      if (typeof category !== "undefined" && filterCategories.length > 0) {
        let _category = filterCategories.filter(x => x.getUrn() === category);

        if (_category[0]) {
          description = _category[0].getDescription();
        }
      } else if (filterCategories.length > 0) {
        description =
          "Полезные статьи для автовладельцев от экспертов: советы по выбору, содержанию, ремонту и продаже автомобилей. ";
      }
    }
    return description;
  };

  /**
   * Формирование ссылок для перехода по страницам в пагринации
   *
   * @param {number} i
   */
  const getPageUrl = i => {
    let url = `/blog/?page=${i}`;
    if (typeof category !== "undefined") {
      url = `${match.url}?page=${i}`;
    }
    return url;
  };

  /**
   * Метод отвечающий за переключении страницы по пагиныции
   *
   * @param {number} i
   */
  const setPageUrl = i => {
    let pageNumberUrl = "";

    if (i > 1) {
      pageNumberUrl = `?page=${i}`;
    }

    if (typeof category !== "undefined") {
      history.push(`${match.url}${pageNumberUrl}`);
    } else {
      history.push(`/blog/${pageNumberUrl}`);
    }

    setReplace(true);
  };

  const handleClickPagination = () => {
    const nextCurrentPage = currentPage + 1;
    if (nextCurrentPage <= postsPagesCount) {
      if (typeof category !== "undefined") {
        history.push(`${match.url}?page=${nextCurrentPage}`);
      } else {
        // history.push(``);
      }
      setCurrentPage(nextCurrentPage);
    } else if (nextCurrentPage > postsPagesCount) {
      setCurrentPage(postsPagesCount);
    }
  };

  const handleClickBlogTag = categoryTag => {
    // history.push();
    setCurrentPage(1);
  };

  if (category) {
    activeCategory = filterCategories.filter(categoryItem => category === categoryItem.getUrn())[0];
  }

  return (
    <>
      <Helmet
        title={getTitle()}
        meta={[
          {
            name: "viewport",
            content: "width=device-width, initial-scale=1",
          },
          {
            name: "format-detection",
            content: "telephone=no",
          },
          { property: "og:title", content: getTitle() },
          { name: "description", content: getDescription() },
        ]}
      />
      <ol className="breadcrumb" itemScope itemType="https://schema.org/BreadcrumbList">
        <li itemProp="itemListElement" itemType="https://schema.org/ListItem" itemScope>
          <a href="/" title="euroauto.ru" itemProp="item">
            <span itemProp="name" className="hidden">
              Главная
            </span>
            <svg
              className="svg-icon"
              xmlnsXlink="http://www.w3.org/1999/xlink"
              width="15px"
              height="15px"
            >
              <use href="#home"></use>
            </svg>
          </a>
          <meta itemProp="position" content="1"></meta>
        </li>
        {!activeCategory && (
          <li
            className="active"
            itemProp="itemListElement"
            itemType="https://schema.org/ListItem"
            itemScope
          >
            <span itemProp="name">Блог</span>
            <meta itemProp="position" content="2"></meta>
          </li>
        )}

        {activeCategory && (
          <>
            <li itemProp="itemListElement" itemType="https://schema.org/ListItem" itemScope>
              <a href="/blog/" itemProp="item">
                <span itemProp="name">Блог</span>
              </a>
              <meta itemProp="position" content="2"></meta>
            </li>
            <li
              className="active"
              itemProp="itemListElement"
              itemType="https://schema.org/ListItem"
              itemScope
            >
              <span itemProp="name">{activeCategory.getLabel()}</span>
              <meta itemProp="position" content="3"></meta>
            </li>
          </>
        )}
      </ol>

      <PageHeader>Блог</PageHeader>

      <FilterPanel
        className="filters"
        filterButtons={filterCategories}
        activeCategory={category}
        clickTag={handleClickBlogTag}
      />

      <Loader notLoadedWhile={!isFetched && isFetching} position={"absolute"} />

      <div className="fx wrap" itemScope itemType="https://schema.org/Blog">
        <React.Fragment>
          <meta
            itemProp="description"
            content="Полезные статьи для автовладельцев от экспертов: советы по выбору, содержанию, ремонту и продаже автомобилей."
          />
          {posts.map(post => (
            <BlogPost
              key={post.getId() + "-" + post.getH1()}
              post={post}
              clickTag={handleClickBlogTag}
            />
          ))}
        </React.Fragment>
      </div>
      {posts.length > 0 && (
        <PaginationBlock
          infinityButtonOnClick={handleClickPagination}
          infinityButtonIsShow={currentPage < postsPagesCount}
          activePage={parseInt(currentPage, 10)}
          totalItemsCount={postsPagesCount * COUNT_PER_PAGE}
          showNumber={true}
          itemsCountPerPage={COUNT_PER_PAGE}
          getPageUrl={getPageUrl}
          setPageUrl={setPageUrl}
        />
      )}
    </>
  );
};

function usePrevious(value) {
  const ref = useRef();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
}
