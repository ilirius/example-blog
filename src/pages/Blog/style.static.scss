#blog_root {
	.wrap {
		align-items: stretch;
		flex-wrap: wrap;
		justify-content: flex-start;
		margin-bottom: 60px;
		position: relative;

		animation: slide-down 0.7s ease;

		@media (min-width: 1159px) {
			& {
				justify-content: flex-start;
			}
		}

		@media (min-width: 320px) and (max-width: 739px) {
			& {
				justify-content: center;
			}
		}
	}

	.filters {
		margin-bottom: 30px;
	}

	.filter {
		display: inline-block;
		margin: 3px 3px 3px 0;

		&-block__item {
			margin: 0;
		}
	}

	.filter.active .filter-item__link{
		background-color: #337ab7;
		color: #ffffff;
	}

	@media (max-width: 992px) {
		.filter-panel {
			margin-top: 0;
			flex-wrap: wrap;
			justify-content: flex-start;
			align-content: flex-start;
			align-items: flex-start;
		}

		// .filter-block__item {
		// 	margin-bottom: 15px;
		// 	margin-right: 15px;
		// }
	}

	@keyframes slide-down {
		0% {
			opacity: 0;
			transform: translateY(20px);
		}

		100% {
			opacity: 1;
			transform: translateY(0);
		}
	}
}
