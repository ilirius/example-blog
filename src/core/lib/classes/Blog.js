import { format } from "date-fns";
import BlogCategory from "./Blog/BlogCategory";

/**
 * Класс работы с блогом.
 *
 * @export
 * @class Blog
 */
export default class Blog {
  _data = {};
  _id = null;

  constructor(id, data) {
    this._data = data;
    this._id = id;
  }

  /**
   * Получение ссылки на основной банер страницы блога.
   * Временно используется для вывода в списке постов.
   *
   * @returns
   * @memberof Blog
   * @returns {string}
   */
  getBannerUrl() {
    /**
     * @type {string}
     */
    let url = this.__get_data_value__("", "");
    return url;
  }

  /**
   * Получение основного текста статьи блога в формате HTML
   *
   * @readonly
   * @memberof Blog
   * @returns {string}
   */
  getContentArticle() {
    return this.__get_data_value__("");
  }

  /**
   * Возвращает дату создания статьи блога. Преобразовывая из 'YYYY-MM-DD' => 'DD.MM.YYYY'
   *
   * @readonly
   * @memberof Blog
   * @returns {string} формат 'DD.MM.YYYY'
   */
  getCreatedAt(data_format = "dd.MM.yyyy") {
    let createdAt = this.__get_data_value__("");

    if (createdAt) {
      createdAt = format(new Date(createdAt), data_format);
    }

    return createdAt;
  }

  /**
   * Возвращает описание страницы статьи для тега <meta name="description" content=""/>
   *
   * @readonly
   * @memberof Blog
   * @returns {string}
   */
  getDescription() {
    return this.__get_data_value__("");
  }

  /**
   * Признак активности статьи
   *
   * @returns
   * @memberof Blog
   * @returns {boolean}
   */
  getEnabled() {
    return this.__get_data_value__("", true);
  }

  /**
   * Возвращает заголовок статьи
   *
   * @readonly
   * @memberof Blog
   * @returns {string}
   */
  getH1() {
    return this.__get_data_value__("");
  }

  /**
   *
   *
   * @readonly
   * @memberof Blog
   * @returns {number}
   */
  getId() {
    return this._id;
  }

  /**
   * Короткий текст статьи
   *
   * @returns
   * @memberof Blog
   * @returns {string}
   */
  getPreviewText() {
    return this.__get_data_value__("");
  }

  /**
   * Возвращает заголовок страницы статьи для тега <title></title>
   *
   * @returns
   * @memberof Blog
   * @returns {string}
   */
  getTitle() {
    return this.__get_data_value__("");
  }

  /**
   *
   *
   * @returns
   * @memberof Blog
   * @returns {string}
   */
  getUrn() {
    return "";
  }

  /**
   *
   *
   * @returns
   * @memberof Blog
   * @returns {BlogCategory}
   */
  getCategory() {
    const category = this.__get_data_value__("", {});
    return new BlogCategory(category.id, category);
  }

  /**
   * Возвращает значение из внутреннего массива данных
   * @param {string} $key
   * @param {any} $default
   * @return {any}
   */
  __get_data_value__($key, $default = null) {
    return this._data[$key] ? this._data[$key] : $default;
  }
}
