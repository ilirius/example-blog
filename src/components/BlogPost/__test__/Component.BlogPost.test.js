import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { BlogPost } from "../BlogPost";
import Blog from "../../../lib/classes/Blog";

const __mock__ = {
  id: 53,
  h1: "Всем выйти из тени!",
  title: "Всем выйти из тени!",
  description: "Как купить подержанный автомобиль без скрученного пробега?",
  content_article: null,
  banner_url: "https://c.radikal.ru/c08/1904/18/fe4e6b48c7cf.jpg",
  enabled: true,
  previewText:
    "Продажа машины в современном мире — это уже давно рутина, бизнес. Вот только конкуренция в нём такая, что ради лишней копейки продавцы зачастую и пробег скрутят, и склеивают одно авто из трех.",
  createdAt: "2019-04-18",
  urn: "2019-04-18-autoteka",
  category: {
    id: 1,
    label: "Общее",
    urn: "common",
    postsCount: null,
  },
};
const post = new Blog(__mock__.id, __mock__);

describe("Components BlogPost", () => {
  const wrapper = shallow(<BlogPost post={post} />);
  it("Рендериться без сбоев", () => {
    let wrapperSnapshot = renderer.create(wrapper).toJSON();
    expect(wrapperSnapshot).toMatchSnapshot();
    wrapperSnapshot = renderer.create(wrapper).toJSON();
    expect(wrapperSnapshot).toMatchSnapshot();
  });

  describe("Метод Render", () => {
    it("Ссылка на страницу блога", () => {
      const node = wrapper.exists("a");
      expect(node).toEqual(true);
    });

    it("Содержит контейнер с картинкой", () => {
      expect(wrapper.find(".img")).toBeTruthy();
    });

    it("Заголовок статьи не пуст", () => {
      expect(wrapper.find(".content_title").text()).toHaveLength(19);
    });

    it("Содержит контейнер для вывода превью блога", () => {
      expect(wrapper.find(".content_text")).toBeTruthy();
    });

    it("Содержит подвал карточки блога", () => {
      expect(wrapper.find(".content_footer")).toBeTruthy();
    });
  });
});
