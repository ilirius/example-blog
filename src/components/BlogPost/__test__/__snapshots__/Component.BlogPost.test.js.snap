// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components BlogPost Рендериться без сбоев 1`] = `
<a
  className="post"
  href="/blog/common/2019-04-18-autoteka"
>
  <div
    className="img"
  >
    <div
      className="lazyload-placeholder"
      style={
        Object {
          "height": 127,
        }
      }
    />
  </div>
  <div
    className="fx content"
  >
    <div
      className="content_title"
    >
      Всем выйти из тени!
    </div>
    <div
      className="content_text"
    >
      Продажа машины в современном мире — это уже давно рутина, бизнес. Вот только конкуренция в нём такая, что ради лишней копейки продавцы зачастую и пробег скрутят, и склеивают одно авто из трех.
    </div>
    <div
      className="fx content_footer"
    >
      <div
        className="content_footer__data"
      >
        18.04.2019
      </div>
      <button
        className="footer__tag content_footer__tag"
        onClick={[Function]}
      >
        Общее
      </button>
    </div>
  </div>
</a>
`;

exports[`Components BlogPost Рендериться без сбоев 2`] = `
<a
  className="post"
  href="/blog/common/2019-04-18-autoteka"
>
  <div
    className="img"
  >
    <div
      className="lazyload-placeholder"
      style={
        Object {
          "height": 127,
        }
      }
    />
  </div>
  <div
    className="fx content"
  >
    <div
      className="content_title"
    >
      Всем выйти из тени!
    </div>
    <div
      className="content_text"
    >
      Продажа машины в современном мире — это уже давно рутина, бизнес. Вот только конкуренция в нём такая, что ради лишней копейки продавцы зачастую и пробег скрутят, и склеивают одно авто из трех.
    </div>
    <div
      className="fx content_footer"
    >
      <div
        className="content_footer__data"
      >
        18.04.2019
      </div>
      <button
        className="footer__tag content_footer__tag"
        onClick={[Function]}
      >
        Общее
      </button>
    </div>
  </div>
</a>
`;
